crates/bevy_render/src/shader/shader_reflect.rs

```diff
@@ -1,300 +1,378 @@
-                has_dynamic_offset: false,
-                property: reflect_uniform(type_description),
-            },
-        ),
-        ReflectDescriptorType::SampledImage => (
-            &binding.name,
-            BindType::Texture {
-                view_dimension: reflect_dimension(type_description),
-                sample_type: TextureSampleType::Float { filterable: true },
-                multisampled: false,
-            },
-        ),
-        ReflectDescriptorType::StorageBuffer => (
-            &type_description.type_name,
-            BindType::StorageBuffer {
-                has_dynamic_offset: false,
-                readonly: true,
-            },
-        ),
-        // TODO: detect comparison "true" case: https://github.com/gpuweb/gpuweb/issues/552
-        // TODO: detect filtering "true" case
-        ReflectDescriptorType::Sampler => (
-            &binding.name,
-            BindType::Sampler {
-                comparison: false,
-                filtering: true,
-            },
-        ),
-        _ => panic!("Unsupported bind type {:?}.", binding.descriptor_type),
-    };
+fn reflect_bind_groups(
+    module: &naga::Module,
+    shader_stage: BindingShaderStage,
+) -> Vec<BindGroupDescriptor> {
+    let mut bind_groups = HashMap::<u32, Vec<_>>::default();
+
+    let binding_descriptors = module
+        .global_variables
+        .iter()
+        .map(|(_, variable)| variable)
+        .filter_map(|variable| {
+            let binding = variable.binding.as_ref()?;
+            Some((variable, binding))
+        })
+        .map(|(variable, binding)| {
+            let name = variable
+                .name
+                .clone()
+                .expect("resource binding without name");
+
+            let ty = module
+                .types
+                .try_get(variable.ty)
+                .expect("resource binding references inexistent type");
 
-    let shader_stage = match shader_stage {
-        ReflectShaderStageFlags::COMPUTE => BindingShaderStage::COMPUTE,
-        ReflectShaderStageFlags::VERTEX => BindingShaderStage::VERTEX,
-        ReflectShaderStageFlags::FRAGMENT => BindingShaderStage::FRAGMENT,
-        _ => panic!("Only one specified shader stage is supported."),
-    };
+            let name = if name.is_empty() {
+                ty.name.clone().unwrap_or_default()
+            } else {
+                name
+            };
+
+            let bind_type = reflect_bind_type(&module, variable.class, binding, &ty.inner);
+            let binding_descriptor = BindingDescriptor {
+                index: binding.binding,
+                bind_type,
+                name,
+                shader_stage,
+            };
+            (binding, binding_descriptor)
+        });
 
-    BindingDescriptor {
-        index: binding.binding,
-        bind_type,
-        name: name.to_string(),
-        shader_stage,
+    for (binding, binding_descriptor) in binding_descriptors {
```

Comment:
```suggestion
    let mut bind_groups = HashMap::<u32, Vec<_>>::default();
    for (binding, binding_descriptor) in binding_descriptors {
```

Declare the `HashMap` (where it is first used) and not at the beginning of the function.